// ngx-popup-dialog/esm2015/lib/ngx-popup-dialog.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { PopupDialog } from './popup-dialog/popup-dialog';
import { PopupDialogService } from './popup-dialog/popup-dialog.service';
import { MatDialogModule } from '@angular/material/dialog';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
export class NgxPopupDialogModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: NgxPopupDialogModule,
            providers: [PopupDialogService]
        };
    }
}
NgxPopupDialogModule.decorators = [
    { type: NgModule, args: [{
                declarations: [PopupDialog],
                imports: [
                    CommonModule,
                    FormsModule,
                    MatDialogModule
                ],
                entryComponents: [PopupDialog],
                exports: []
            },] }
];
