// ngx-popup-dialog/esm2015/lib/popup-dialog/popup-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Overlay } from '@angular/cdk/overlay';
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { PopupDialog } from './popup-dialog';
/**
 * @record
 */
export function PopupDialogServiceConfig() { }
if (false) {
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.direction;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.coverTriggeringElement;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.scaleToTopOnBottomOverflow;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.maxWidth;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.maxHeight;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.suppressCloseOnClickSelectors;
}
/**
 * @record
 */
export function PopupDialogConfig() { }
if (false) {
    /** @type {?|undefined} */
    PopupDialogConfig.prototype.data;
}
export class PopupDialogService {
    /**
     * @param {?} _overlay
     * @param {?} _dialogService
     */
    constructor(_overlay, _dialogService) {
        this._overlay = _overlay;
        this._dialogService = _dialogService;
    }
    /**
     * @template T
     * @param {?} componentOrTemplateRef
     * @param {?} triggeringElement
     * @param {?=} config
     * @return {?}
     */
    open(componentOrTemplateRef, triggeringElement, config) {
        if (!config)
            config = {};
        /** @type {?} */
        let data = config.data;
        delete config.data;
        /** @type {?} */
        const popupDialogData = {
            config: Object.assign({}, config, { component: componentOrTemplateRef, triggeringElement: triggeringElement, maxWidth: (typeof config.maxWidth === "number") ? config.maxWidth + 'px' : config.maxWidth, maxHeight: (typeof config.maxHeight === "number") ? config.maxHeight + 'px' : config.maxHeight, scaleToTopOnBottomOverflow: config.scaleToTopOnBottomOverflow == null ? true : config.scaleToTopOnBottomOverflow, suppressCloseOnClickSelectors: typeof config.suppressCloseOnClickSelectors === "string" ? [config.suppressCloseOnClickSelectors] : config.suppressCloseOnClickSelectors }),
            data: data
        };
        /** @type {?} */
        const popupDialogConfig = {
            backdropClass: 'cdk-overlay-transparent-backdrop',
            hasBackdrop: false,
            scrollStrategy: this._overlay.scrollStrategies.noop(),
            panelClass: 'ngx-popup-dialog-overlay',
            restoreFocus: false,
            data: popupDialogData
        };
        /** @type {?} */
        let dialogRef = this._dialogService.open(PopupDialog, popupDialogConfig);
        return dialogRef;
    }
}
PopupDialogService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
PopupDialogService.ctorParameters = () => [
    { type: Overlay },
    { type: MatDialog }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    PopupDialogService.prototype._overlay;
    /**
     * @type {?}
     * @private
     */
    PopupDialogService.prototype._dialogService;
}
