// ngx-popup-dialog/esm5/lib/popup-dialog/popup-dialog.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Overlay } from '@angular/cdk/overlay';
import { Injectable } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { PopupDialog } from './popup-dialog';
/**
 * @record
 */
export function PopupDialogServiceConfig() { }
if (false) {
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.direction;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.coverTriggeringElement;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.scaleToTopOnBottomOverflow;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.maxWidth;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.maxHeight;
    /** @type {?|undefined} */
    PopupDialogServiceConfig.prototype.suppressCloseOnClickSelectors;
}
/**
 * @record
 */
export function PopupDialogConfig() { }
if (false) {
    /** @type {?|undefined} */
    PopupDialogConfig.prototype.data;
}
var PopupDialogService = /** @class */ (function () {
    function PopupDialogService(_overlay, _dialogService) {
        this._overlay = _overlay;
        this._dialogService = _dialogService;
    }
    /**
     * @template T
     * @param {?} componentOrTemplateRef
     * @param {?} triggeringElement
     * @param {?=} config
     * @return {?}
     */
    PopupDialogService.prototype.open = /**
     * @template T
     * @param {?} componentOrTemplateRef
     * @param {?} triggeringElement
     * @param {?=} config
     * @return {?}
     */
    function (componentOrTemplateRef, triggeringElement, config) {
        if (!config)
            config = {};
        /** @type {?} */
        var data = config.data;
        delete config.data;
        /** @type {?} */
        var popupDialogData = {
            config: tslib_1.__assign({}, config, { component: componentOrTemplateRef, triggeringElement: triggeringElement, maxWidth: (typeof config.maxWidth === "number") ? config.maxWidth + 'px' : config.maxWidth, maxHeight: (typeof config.maxHeight === "number") ? config.maxHeight + 'px' : config.maxHeight, scaleToTopOnBottomOverflow: config.scaleToTopOnBottomOverflow == null ? true : config.scaleToTopOnBottomOverflow, suppressCloseOnClickSelectors: typeof config.suppressCloseOnClickSelectors === "string" ? [config.suppressCloseOnClickSelectors] : config.suppressCloseOnClickSelectors }),
            data: data
        };
        /** @type {?} */
        var popupDialogConfig = {
            backdropClass: 'cdk-overlay-transparent-backdrop',
            hasBackdrop: false,
            scrollStrategy: this._overlay.scrollStrategies.noop(),
            panelClass: 'ngx-popup-dialog-overlay',
            restoreFocus: false,
            data: popupDialogData
        };
        /** @type {?} */
        var dialogRef = this._dialogService.open(PopupDialog, popupDialogConfig);
        return dialogRef;
    };
    PopupDialogService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    PopupDialogService.ctorParameters = function () { return [
        { type: Overlay },
        { type: MatDialog }
    ]; };
    return PopupDialogService;
}());
export { PopupDialogService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PopupDialogService.prototype._overlay;
    /**
     * @type {?}
     * @private
     */
    PopupDialogService.prototype._dialogService;
}
